### DISCUSSION TITLE
Stand alone impute vs. na.impute argument

### DISCUSSION THREAD

**Discussion by cmnelson1** [kogalur/randomForestSRC#322] [Q&A]
Thank you for the randomForestSRC package and the github website!

We are running a Random Survival Forest with a sample of 801 participants and 67 features.  We have some missing data and want to impute missing values using the impute function in randomForest SRC (in this case, we want to use RSF to get the VIMPs for variable selection to run in another method and compare results with RSF). In the past, I have imputed missing values among the features by means of unsupervised splitting AND using a data set where I removed the outcome variable from the dataset.  I have used this approach both for running a RF on the imputed data set, and also using the imputed data set in another method such as boosting.  Logically, it makes sense to me to remove the outcome variable so that the imputing is unrelated to our outcome. But, in many of the examples in the SRC manual, even that for unsupervised splitting, it looks like the full data set is being used for the impute.  My questions are:
 
1)	For imputation only mode, am I correct that it is best to impute the data set after removing the outcome variable from the data set (if the goal is to then analyze with respect to that outcome)?
2)	In terms of best practices for the stand-alone random survival forest, is it best to pre-impute the data set (again, first removing the outcome variable) and then run random survival forest on this already imputed data after reuniting it with the outcome variable OR using the na.impute argument?
3)	Is it best of all to use the mf.q argument on a stand-alone impute on the data set prior to analysis (again, omitting the outcome variable to do so)?
 
 Any advice would be much appreciated!  Thank you!

**Comment by ishwaran** [comment]
1. Yes.
2. Generally, it's better to pre-impute the data (without the response/outcome) and then run RSF on imputed data.  It is also much faster than using OTF (on the fly imputation).
3. We found that missForest and the multivariate missForest (`mf.q `argument) had the highest imputation accuracy.

Generally speaking it's better to pre-impute the data. The advantage of OTF, where data is imputed as tree is grown, is that you can use it on test data in the future which is unlike a method like missForest where you pre-impute the data up front. So from a ML perspective it's nice that you can handle prediction on future data without worrying about missing data.  On the other hand, the `predict.rfsrc` has a new option `na.action='na.random'` which can handle missing data without the need for OTF.  So maybe OTF relevance is becoming less over time.

**Reply by cmnelson1** [reply]
Thank you Dr. Ishwaran for your responses to my questions.  Very helpful!
